docs/src/design/changing_the_primal.md

```diff
@@ -0,0 +1,516 @@
+Conceptually the `PullbackMemory` is a fixed thing it the contents of the tape for a particular operation.
+It is fully determined by the end of the primal pass.
+The during the gradient (reverse) pass the `PullbackMemory` is used to successively compute the `ȳ`  argument.
+So it makes sense to have `PullbackMemory` being a callable object that acts on the sensitivity.
+We can do that via call overloading:
+```julia
+y = f(x)  # primal program
+y, pb = augmented_primal(f, x)
+x̄ = pb(ȳ)
+```
+
+```@raw html
+<details><summary>Example for `sin`</summary>
+```
+```julia
+function augmented_primal(::typeof(sin), x)
+  y, cx = sincos(x)
+  return y, PullbackMemory(sin; cx=cx)
+end
+(pb::PullbackMemory)(ȳ) = ȳ * pb.cx
+```
+
+```@raw html
+</details>
+```
+
+```@raw html
+<details><summary>Example for the logistic sigmoid</summary>
+```
+```julia
+function augmented_primal(::typeof(σ), x)
+  ex = exp(x)
+  y = ex/(1 + ex)
+  return y, PullbackMemory(σ; y=y, ex=ex)
+end
+
+(pb::PullbackMemory{typeof(σ)})(ȳ) = ȳ * pb.y/(1 + pb.ex)
+```
+```@raw html
+</details>
+```
+
+Those looking closely will spot what we have done here.
+We now have an object (`pb`) that acts on the cotangent of the output of the primal (`ȳ`) to give us the cotangent of the input of the primal function (`x̄`).
+_`pb` is not just the **memory** of state required for the `pullback`, it **is** the pullback._
+
+We have one final thing to do.
+Let's think about making the code easy to modify.
+Let's go back and think about the changes we would have make to go from our original way of writing that only used the inputs/outputs, to one that used the intermediate state.
+
+```@raw html
+<details><summary>Example for `sin`</summary>
+```
+To rewrite that original formulation in the new pullback form we have:
+```julia
+function augmented_primal(::typeof(sin), x)
+  y = sin(x)
+  return y, PullbackMemory(sin; x=x)
+end
+(pb::PullbackMemory)(ȳ) = ȳ * cos(pb.x)
+```
+To go from that to:
+```julia
+function augmented_primal(::typeof(sin), x)
+  y, cx = sincos(x)
+  return y, PullbackMemory(sin; cx=cx)
+end
+(pb::PullbackMemory)(ȳ) = ȳ * pb.cx
+```
+```@raw html
+</details>
+```
+
+```@raw html
+<details><summary>Example for the logistic sigmoid</summary>
+```
+```julia
+function augmented_primal(::typeof(σ), x)
+  y = σ(x)
+  return y, PullbackMemory(σ; y=y, x=x)
+end
+(pb::PullbackMemory{typeof(σ)})(ȳ) = ȳ * pb.y * σ(-pb.x)
+```
+to get to:
+```julia
+function augmented_primal(::typeof(σ), x)
+  ex = exp(x)
+  y = ex/(1 + ex)
+  return y, PullbackMemory(σ; y=y, ex=ex)
+end
+(pb::PullbackMemory{typeof(σ)})(ȳ) = ȳ * pb.y/(1 + pb.ex)
+```
+```@raw html
+</details>
+```
+
+(NB: there is actually a further optimization that can be made to the logistic sigmoid, to avoid remembering two things and just remember one.
+As an exercise to the reader, consider how the code would need to be changed and where.)
```

Comment:
"exercise to the reader" is a pet peeve of mine.

it sounds such a smart-arse thing to say. At least put the answer in a footnote. Perferably ditch the "exercise to the reader" and just tell us there is a further optimization and put the details in a footnote.

```suggestion
As an aside, there a further optimization that can be made to the logistic sigmoid, to avoid remembering two things and just remember one, but we'll leave the example simple for now.[^FOOTNOTE]
```

Reply:
Done